File: packages/prime-agent/src/project-config.ts
// Root-only project customization (no shared-pi import per ADR-CORE-014
// isolation). Workflow then rules; absent/empty skipped, unusable throws
// rel-only diagnostics (see ADR-CORE-006).

import { lstatSync, readFileSync, realpathSync } from 'node:fs';
import path from 'node:path';

export const PROJECT_WORKFLOW_REL = '.maestria/workflow.md';
export const PROJECT_RULES_REL = '.maestria/rules.md';
/** Deterministic load order: workflow sequencing first, then project rules. */
export const PROJECT_CONFIG_REL_PATHS = [PROJECT_WORKFLOW_REL, PROJECT_RULES_REL] as const;

export type ProjectEntryKind = 'missing' | 'file' | 'directory' | 'other';

export interface ProjectSection {
  content: string;
  rel: string;
}

export interface ProjectConfigFs {
  kindOf: (candidate: string) => ProjectEntryKind;
  readFile: (candidate: string) => string;
  resolveLink: (candidate: string) => string;
}

const isNonEmpty = (value: unknown): value is string => typeof value === 'string' && value !== '';

const isEnoent = (error: unknown): boolean =>
  typeof error === 'object' &&
  error !== null &&
  'code' in error &&
  (error as { code?: unknown }).code === 'ENOENT';

const isSanitizedDiagnostic = (error: unknown): error is Error =>
  error instanceof Error && error.message.startsWith('[maestria] Project config');

/** Raw fs failures name only the rel file and kind; sanitized ones pass through. */
const sanitizeFsError = (rel: string, fallback: string, error: unknown): Error => {
  if (isSanitizedDiagnostic(error)) {
    return error;
  }
  return new Error(`[maestria] Project config "${rel}" ${fallback}`);
};

/** Reject a present-but-unusable entry with a rel-only diagnostic. */
const assertFileKind = (rel: string, kind: ProjectEntryKind): void => {
  if (kind === 'directory') {
    throw new Error(`[maestria] Project config "${rel}" is a directory, expected a file`);
  }
  if (kind === 'other') {
    throw new Error(`[maestria] Project config "${rel}" is not a regular file`);
  }
};

const defaultFs: ProjectConfigFs = {
  kindOf: (candidate) => {
    try {
      const stat = lstatSync(candidate);
      if (stat.isDirectory()) {
        return 'directory';
      }
      if (stat.isFile() || stat.isSymbolicLink()) {
        return 'file';
      }
      return 'other';
    } catch (error) {
      if (isEnoent(error)) {
        return 'missing';
      }
      throw error;
    }
  },
  readFile: (candidate) => readFileSync(candidate, 'utf-8'),
  resolveLink: (candidate) => realpathSync(candidate),
};

const escapesRoot = (root: string, resolved: string): boolean => {
  const relative = path.relative(root, resolved);
  return relative === '..' || relative.startsWith(`..${path.sep}`) || path.isAbsolute(relative);
};

/** Load one file: missing/empty yields nothing, unusable throws rel-only. */
const loadOneSection = (
  normalizedRoot: string,
  rel: string,
  fs: ProjectConfigFs,
): ProjectSection | undefined => {
  const candidate = path.join(normalizedRoot, rel);
  let kind: ProjectEntryKind;
  try {
    kind = fs.kindOf(candidate);
  } catch (error) {
    throw sanitizeFsError(rel, 'cannot be accessed', error);
  }
  if (kind === 'missing') {
    return undefined;
  }
  assertFileKind(rel, kind);

  let resolved: string;
  try {
    resolved = fs.resolveLink(candidate);
  } catch (error) {
    throw sanitizeFsError(rel, 'cannot be resolved', error);
  }
  if (escapesRoot(normalizedRoot, resolved)) {
    throw new Error(`[maestria] Project config "${rel}" resolves outside the project root`);
  }
  let targetKind: ProjectEntryKind;
  try {
    targetKind = fs.kindOf(resolved);
  } catch (error) {
    throw sanitizeFsError(rel, 'cannot be accessed', error);
  }
  if (targetKind === 'missing') {
    throw new Error(`[maestria] Project config "${rel}" cannot be accessed`);
  }
  assertFileKind(rel, targetKind);
  let content: string;
  try {
    content = fs.readFile(candidate);
  } catch (error) {
    throw sanitizeFsError(rel, 'exists but cannot be read', error);
  }
  if (content === '') {
    return undefined;
  }
  return { content, rel };
};

/** Read customization in order; missing/empty skipped, unusable throws. */
export const loadProjectSections = (
  root: string,
  fs: ProjectConfigFs = defaultFs,
): ProjectSection[] => {
  if (!isNonEmpty(root)) {
    return [];
  }
  let normalizedRoot: string;
  try {
    normalizedRoot = realpathSync(root);
  } catch (error) {
    if (!isEnoent(error)) {
      // oxlint-disable-next-line preserve-caught-error -- the diagnostics contract forbids attaching the raw error (absolute paths may leak through host error serialization); the kind is preserved in the message.
      throw new Error('[maestria] Project config root cannot be accessed');
    }
    // Absent root: keep the lexical path so the kind check below reports
    // each file missing (normal, skipped) instead of erroring.
    normalizedRoot = path.resolve(root);
  }
  const sections: ProjectSection[] = [];

  for (const rel of PROJECT_CONFIG_REL_PATHS) {
    const section = loadOneSection(normalizedRoot, rel, fs);
    if (section !== undefined) {
      sections.push(section);
    }
  }

  return sections;
};

/** Format one section; the header keeps subordinate status visible. */
export const formatProjectSection = (section: ProjectSection): string =>
  [
    `Project customization from ${section.rel} (subordinate guidance: it may replace configurable workflows but never waives safety, authorization, or host permissions):`,
    section.content,
  ].join('\n');

/** Fail-loud banner with STOP; callers surface it via notify plus prompt. */
export const formatProjectErrorBanner = (message: string): string =>
  [
    'Project customization failed to load. STOP: do not execute the user request on potentially overridden configuration.',
    'Report this error to the user and wait for the project files to be fixed.',
    message,
  ].join('\n');

File: packages/hermes/src/maestria_hermes/project_config.py
"""Project-root customization loader for the maestria Hermes plugin.

Root-only workflow then rules, fresh read every turn; absent/empty leaves
the context unchanged, present-but-unusable raises ProjectConfigError with
rel path and kind only. See ADR-CORE-006 and docs/runtime-support-matrix.md.
"""

from __future__ import annotations

import logging
import os
import stat
from dataclasses import dataclass
from pathlib import Path
from typing import Callable

logger = logging.getLogger(__name__)

PROJECT_WORKFLOW_REL = ".maestria/workflow.md"
PROJECT_RULES_REL = ".maestria/rules.md"

# Workflow first, then rules, matching the canonical orchestrator guidance.
PROJECT_CONFIG_REL_PATHS = (PROJECT_WORKFLOW_REL, PROJECT_RULES_REL)


class ProjectConfigError(Exception):
    """A project file is present but unusable."""


@dataclass(frozen=True)
class ProjectSection:
    """One loaded project file with its root-relative path."""

    content: str
    rel: str


def get_project_root() -> str | None:
    """Return the session root for this call, or None when absent."""
    try:
        return os.getcwd() or None
    except OSError:
        logger.debug("maestria project root unavailable (cwd unreadable)")
        return None


def _entry_kind(candidate: str) -> str:
    """Classify without following symlinks: missing, file, directory, other."""
    try:
        st = os.lstat(candidate)
    except FileNotFoundError:
        return "missing"
    except OSError:
        raise
    mode = st.st_mode
    if stat.S_ISDIR(mode):
        return "directory"
    if stat.S_ISREG(mode) or stat.S_ISLNK(mode):
        return "file"
    return "other"


def _resolve_link(candidate: str, rel: str) -> str:
    """Strictly resolve symlinks or raise visibly."""
    try:
        return str(Path(candidate).resolve(strict=True))
    except (OSError, RuntimeError) as exc:
        raise ProjectConfigError(
            f'[maestria] Project config "{rel}" cannot be resolved'
        ) from exc


def _escapes_root(root: str, resolved: str) -> bool:
    """Return True when *resolved* lies outside *root*."""
    relative = os.path.relpath(resolved, root)
    return (
        relative == os.pardir
        or relative.startswith(os.pardir + os.sep)
        or os.path.isabs(relative)
    )


def _read_text(candidate: str, rel: str) -> str:
    """Read UTF-8 text or raise visibly without leaking content."""
    try:
        with open(candidate, encoding="utf-8") as handle:
            return handle.read()
    except (OSError, UnicodeDecodeError) as exc:
        raise ProjectConfigError(
            f'[maestria] Project config "{rel}" exists but cannot be read'
        ) from exc


def _require_file_kind(rel: str, kind: str) -> None:
    """Raise rel-only when a present entry is not a regular file."""
    if kind == "file":
        return
    suffix = (
        "is a directory, expected a file" if kind == "directory" else "is not a regular file"
    )
    raise ProjectConfigError(f'[maestria] Project config "{rel}" {suffix}')


def load_project_sections(
    root: str,
    *,
    kind_of: Callable[[str], str] = _entry_kind,
    read_file: Callable[[str, str], str] = _read_text,
    resolve_link: Callable[[str, str], str] = _resolve_link,
) -> list[ProjectSection]:
    """Load sections in contract order; unusable entries raise rel-only."""
    if not isinstance(root, str) or not root:
        return []
    normalized_root = os.path.realpath(root)
    sections: list[ProjectSection] = []

    for rel in PROJECT_CONFIG_REL_PATHS:
        candidate = os.path.join(normalized_root, rel)
        try:
            kind = kind_of(candidate)
        except ProjectConfigError:
            raise
        except (OSError, RuntimeError) as exc:
            raise ProjectConfigError(
                f'[maestria] Project config "{rel}" cannot be accessed'
            ) from exc
        if kind == "missing":
            continue
        _require_file_kind(rel, kind)

        try:
            resolved = resolve_link(candidate, rel)
        except ProjectConfigError:
            raise
        except (OSError, RuntimeError) as exc:
            raise ProjectConfigError(
                f'[maestria] Project config "{rel}" cannot be resolved'
            ) from exc
        if _escapes_root(normalized_root, resolved):
            raise ProjectConfigError(
                f'[maestria] Project config "{rel}" resolves outside the project root'
            )
        # Resolved targets are rechecked before reading so FIFOs and
        # special files fail here instead of blocking or misreading.
        try:
            target_mode = os.stat(resolved).st_mode
        except OSError as exc:
            raise ProjectConfigError(
                f'[maestria] Project config "{rel}" cannot be accessed'
            ) from exc
        if stat.S_ISDIR(target_mode):
            raise ProjectConfigError(
                f'[maestria] Project config "{rel}" is a directory, expected a file'
            )
        if not stat.S_ISREG(target_mode):
            raise ProjectConfigError(
                f'[maestria] Project config "{rel}" is not a regular file'
            )
        try:
            content = read_file(candidate, rel)
        except ProjectConfigError:
            raise
        except (OSError, UnicodeDecodeError, RuntimeError) as exc:
            raise ProjectConfigError(
                f'[maestria] Project config "{rel}" exists but cannot be read'
            ) from exc
        if content != "":
            sections.append(ProjectSection(content=content, rel=rel))

    return sections


def format_project_section(section: ProjectSection) -> str:
    """Format one section; the header keeps subordinate status visible."""
    return (
        f"Project customization from {section.rel} (subordinate guidance: "
        f"it may replace configurable workflows but never waives safety, "
        f"authorization, or host permissions):\n{section.content}"
    )


def format_project_error(message: str) -> str:
    """Format a fail-open STOP banner; advisory, never enforcement."""
    return (
        "[MAESTRIA PROJECT CONFIG ERROR] "
        f"{message}. STOP: do not run with potentially overridden "
        "configuration; report this error and wait for the file to be "
        "fixed. Project guidance never grants capability or executes."
    )


def build_project_context() -> str:
    """Load and format context for this turn, or "" when absent."""
    root = get_project_root()
    if root is None:
        return ""
    try:
        sections = load_project_sections(root)
    except ProjectConfigError as exc:
        logger.warning("maestria project customization unavailable: %s", exc)
        return format_project_error(str(exc))
    return "\n\n".join(format_project_section(section) for section in sections)
